packages/patternfly-4/react-core/src/components/Form/FormGroup.js

```diff
@@ -0,0 +1,121 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { FormContext } from '../Form/FormContext';
+import { css, getModifier } from '@patternfly/react-styles';
+import styles from '@patternfly/patternfly-next/components/Form/styles.css';
+
+const propTypes = {
+  /** FormGroup field. */
+  children: PropTypes.node,
+  /** Custom class name. */
+  className: PropTypes.string,
+  /** Label text before the field. */
+  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** Sets the FormGroup required. */
+  required: PropTypes.bool,
+  /** Sets the FormGroup invalid. */
+  invalid: PropTypes.bool,
+  /** Helper text after the field. It can be a simple text, a function or an object. */
+  helpertext: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** Helper text after the field when the field is invalid. It can be a simple text, a function or an object. */
+  helpertextInvalid: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** ID of the included field. It has to be the same for proper working. */
+  fieldId: PropTypes.string
+};
+
+const defaultProps = {
+  children: null,
+  className: '',
+  label: undefined,
+  required: false,
+  invalid: false,
+  helpertext: undefined,
+  helpertextInvalid: undefined,
+  fieldId: undefined
+};
+
+const renderHelperText = ({ text, classes, fieldId }) => {
+  if (typeof text === 'string') {
+    return (
+      <p className={classes} id={`${String(fieldId)}-helper`} aria-live="polite">
+        {text}
+      </p>
+    );
+  } else if (typeof text === 'function') {
+    return text();
+  } else if (typeof text === 'object') {
+    return text;
+  }
+  return null;
+};
+
+const renderLabel = ({ label, required, fieldId }) => {
+  const classesRequired = css(styles.formLabelRequired);
+  const classesLabel = css(styles.formLabel);
+
+  if (typeof label === 'string') {
+    if (required === true) {
```

Comment:
I am wondering if this is necessary.  Is there a use case where someone would want to have the label be something other than a string?  This label is also used for accessibility purposes, if it is a node, I don't know how that will work.

Reply:
That is actually desired, given that you may have something really complex as part of Form value you also want to have something complex for label. For instance autogenerated label (thus function) or if someone would want to have for instance icon as a label (can't think of a great reason why, but we shouldn't force consumers one way or another).
It's somewhat of a standard to provide consumer as much freedom of a choice as possible. Of course provide some default features but allow some complex usecases as well.

Reply:
@tlabaj This has become a standard approach in component libraries in React. The most common reasons for this are:
- Rendering icons instead of strings
- Rendering images
- Dynamically changing label value without re-rendering whole form
- Complete styling customization for non-standard use cases.

Reply:
I am not disagreeing with the implementation.  It just differs from what I see in core.  My interpretation of what I see in core is that they expect the label to be a string. I think we need to at least have a discussion with design and css before diverging from what is in core.  :)